planner/core/rule_max_min_eliminate.go

```diff
@@ -18,71 +18,188 @@ import (
-			return
+		return a.checkColCanUseIndex(p.children[0], col)
+	case *DataSource:
+		// Check whether there is an accessPath can use index for col.
+		for _, path := range p.possibleAccessPaths {
+			if path.isTablePath {
+				if col == p.handleCol {
+					return true
+				}
+			} else {
+				if col == path.fullIdxCols[0] && path.fullIdxColLens[0] == types.UnspecifiedLength {
+					return true
+				}
+			}
 		}
+		return false
+	default:
+		return false
+	}
+}
 
-		child := p.Children()[0]
-		ctx := p.SCtx()
-
-		// If there's no column in f.GetArgs()[0], we still need limit and read data from real table because the result should NULL if the below is empty.
-		if len(expression.ExtractColumns(f.Args[0])) > 0 {
-			// If it can be NULL, we need to filter NULL out first.
-			if !mysql.HasNotNullFlag(f.Args[0].GetType().Flag) {
-				sel := LogicalSelection{}.Init(ctx)
-				isNullFunc := expression.NewFunctionInternal(ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), f.Args[0])
-				notNullFunc := expression.NewFunctionInternal(ctx, ast.UnaryNot, types.NewFieldType(mysql.TypeTiny), isNullFunc)
-				sel.Conditions = []expression.Expression{notNullFunc}
-				sel.SetChildren(p.Children()[0])
-				child = sel
-			}
+// cloneSubPlans clones the subPlan. We only consider `Selection` and `DataSource` here,
+// because we have restricted the subPlan in `checkColCanUseIndex`.
+func (a *maxMinEliminator) cloneSubPlans(plan LogicalPlan) LogicalPlan {
+	switch p := plan.(type) {
+	case *LogicalSelection:
+		conditions := make([]expression.Expression, 0, len(p.Conditions))
+		for _, cond := range p.Conditions {
+			conditions = append(conditions, cond.Clone())
+		}
+		sel := LogicalSelection{Conditions: conditions}.Init(p.ctx)
+		sel.SetChildren(a.cloneSubPlans(p.children[0]))
+		return sel
+	case *DataSource:
+		// This is a shallow clone which may not be safe.
+		newDs := *p
+		newDs.self = &newDs
+		return &newDs
+	}
+	// This won't happen, because we have checked the subtree.
+	return nil
+}
 
-			// Add Sort and Limit operators.
-			// For max function, the sort order should be desc.
-			desc := f.Name == ast.AggFuncMax
-			// Compose Sort operator.
-			sort := LogicalSort{}.Init(ctx)
-			sort.ByItems = append(sort.ByItems, &ByItems{f.Args[0], desc})
-			sort.SetChildren(child)
-			child = sort
+// splitAggFuncAndCheckIndices splits the agg to multiple aggs and check whether each agg needs a sort
+// after the transformation. For example, we firstly split the sql: `select max(a), min(a), max(b) from t` ->
+// `select max(a) from t` + `select min(a) from t` + `select max(b) from t`.
+// Then we check whether `a` and `b` have indices. If any of the used column has no index, we cannot eliminate
+// this aggregation.
+func (a *maxMinEliminator) splitAggFuncAndCheckIndices(agg *LogicalAggregation) (aggs []*LogicalAggregation, canEliminate bool) {
+	aggs = make([]*LogicalAggregation, 0, len(agg.AggFuncs))
```

Comment:
It's better to declare this variable after all the validations are passed. So that we can save some memory allocation when the validation result is negative. Which can help us improve the performance in the OLTP scenario whose QPS is typically very high.